src/Web/Slack/Experimental/Views.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | Publishing views using the Blocks API.
+--
+-- This is used for App Home and modals.
+--
+-- <https://api.slack.com/methods/views.publish>
+module Web.Slack.Experimental.Views (
+  -- * Types
+  SlackView (..),
+  HomeTabView (..),
+  ModalView (..),
+  Expected (..),
+
+  -- * Requests and responses
+  PublishReq (..),
+  PublishResp (..),
+  Api,
+  viewsPublish,
+) where
+
+import Data.Aeson qualified as A
+import Data.Aeson.KeyMap qualified as KM
+import Servant.API (AuthProtect, FormUrlEncoded, JSON, Post, ReqBody, (:>))
+import Servant.Client (ClientM, client)
+import Servant.Client.Core (AuthenticatedRequest)
+import Web.FormUrlEncoded (ToForm (..))
+import Web.HttpApiData (ToHttpApiData (..))
+import Web.Slack.AesonUtils (Expected (..), snakeCaseOptions, snakeCaseOptionsEatTrailing)
+import Web.Slack.Experimental.Blocks qualified as B
+import Web.Slack.Experimental.Blocks.Types (SlackPlainTextOnly)
+import Web.Slack.Internal (ResponseJSON (..), SlackConfig (..), mkSlackAuthenticateReq, run)
+import Web.Slack.Pager (Response)
+import Web.Slack.Prelude
+import Web.Slack.Types (UserId)
+
+-- | View definition for some Slack surface. Has an inner type of either
+-- 'ModalView' or 'HomeTabView'.
+--
+-- <https://api.slack.com/reference/surfaces/views>
+data SlackView inner = SlackView
+  { blocks :: Vector B.SlackBlock
+  , privateMetadata :: Maybe Text
+  , callbackId :: Maybe Text
+  , externalId :: Maybe Text
+  , inner :: inner
+  }
+  deriving stock (Show, Eq, Generic)
+
+instance (FromJSON inner) => FromJSON (SlackView inner) where
+  parseJSON = withObject "SlackView" \o -> do
+    blocks <- o .: "blocks"
+    privateMetadata <- o .:? "private_metadata"
+    callbackId <- o .:? "callback_id"
+    externalId <- o .:? "external_id"
+    inner <- parseJSON @inner (A.Object o)
+    pure SlackView {..}
+
+instance (ToJSON inner) => ToJSON (SlackView inner) where
+  toJSON SlackView {..} = case toJSON (inner) of
+    A.Object innerObj ->
+      A.Object $
+        KM.fromList
+          [ "blocks" .= blocks
+          , "private_metadata" .= privateMetadata
+          , "callback_id" .= callbackId
+          , "external_id" .= externalId
+          ]
+          <> innerObj
+    _ -> error "inner of SlackView is not an object"
+
+data HomeTabView = HomeTabView
+  { type_ :: Expected "home"
+  }
+  deriving stock (Show, Eq, Generic)
+
+$(deriveJSON snakeCaseOptionsEatTrailing ''HomeTabView)
+
+-- | Modal view
+--
+-- <https://api.slack.com/reference/surfaces/views#modal>
+data ModalView = ModalView
+  { type_ :: Expected "modal"
+  , title :: SlackPlainTextOnly
+  -- ^ Title of the modal on the top left. Maximum length of 24 characters.
```

Comment:
We could use one of our length-limited text types here, but I doubt it's worth it.